File: WebClient SDK/src/default/ui/openCardInViewUIExtension.ts
import {
  AdvancedCardDialogManager,
  CardUIExtension,
  ICardModel,
  ICardUIExtensionContext
} from 'tessa/ui/cards';
import { executeExtensions, TypeExtensionContext } from 'tessa/cards';
import { TypeSettingsSealed } from 'tessa/cards/types';
import { showError, showLoadingOverlay, tryGetFromSettings, UIContext } from 'tessa/ui';
import { ViewControlViewModel } from 'tessa/ui/cards/controls';
import { ViewMetadataSealed } from 'tessa/views/metadata';
import { DoubleClickInfo } from 'tessa/ui/views';
import { Guid } from 'tessa/platform';
import { openCard } from 'tessa/ui/uiHost';
import { DefaultFormTabWithTasksViewModel } from 'tessa/ui/cards/forms';
import { LocalizationManager } from 'tessa/localization';
import { DefaultCardTypeExtensionTypes } from '../cards/defaultCardTypeExtensionTypes';
import { CardViewControlInfo } from '../cards/cardViewControlInfo';

/**
 * @description
 * UI расширение, реализующее функционал - открыть карточку из представления.
 */
export class OpenCardInViewUIExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext): Promise<void> {
    const result = await executeExtensions(
      DefaultCardTypeExtensionTypes.openCardInView,
      context.card,
      context.model.generalMetadata,
      this.executeInitializedActionAsync,
      context
    );

    context.validationResult.add(result);
  }

  private executeInitializedActionAsync = async (context: TypeExtensionContext): Promise<void> => {
    const extensionContext = context.externalContext as ICardUIExtensionContext;
    const settings = context.settings;
    if (!context.cardTask) {
      await this.attachDoubleClickHandlerAsync(extensionContext.model, settings);
    } else {
      const model = extensionContext.model;
      const tasks = (model.mainForm as DefaultFormTabWithTasksViewModel).tasks;
      if (!tasks) {
        return;
      }
      const task = tasks.find(x => x.taskModel.cardTask === context.cardTask);
      if (task) {
        task.modifyWorkspace(async () => {
          await this.attachDoubleClickHandlerAsync(task.taskModel, settings);
        });
      }
    }
  };

  private async attachDoubleClickHandlerAsync(
    cardModel: ICardModel,
    settings: TypeSettingsSealed | null
  ) {
    const viewControlName = tryGetFromSettings<string>(settings, 'ViewControlAlias', '');
    const viewModel = cardModel.controls.get(viewControlName) as ViewControlViewModel;
    if (!viewModel) {
      return;
    }
    const prefixReference = tryGetFromSettings<string>(settings, 'ViewReferencePrefix', '')?.trim();
    const mapping = OpenCardInViewUIExtension.tryGetCardReferenceMapping(
      viewModel,
      prefixReference
    );
    const displayInDialog = tryGetFromSettings<boolean>(settings, 'IsOpenCardInDialog', false);
    const dialogName = tryGetFromSettings<string>(settings, 'CardDialogName', '');
    viewModel.doubleClickAction = async (info: DoubleClickInfo) => {
      if (prefixReference && !mapping) {
        showError(
          LocalizationManager.instance.format(
            '$UI_Cards_TypesEditor_Exception_RefSectionInView',
            prefixReference,
            viewModel.cardTypeControl.caption,
            viewModel.block.cardTypeBlock.caption,
            viewModel.block.form.cardTypeForm['tabCaption'],
            viewModel.cardModel.cardType.caption
          )
        );
      }
      await this.doubleClickHandlerAsync(info, mapping, displayInDialog, dialogName);
    };
  }

  private async doubleClickHandlerAsync(
    info: DoubleClickInfo,
    mapping: { entityID: string } | null,
    displayInDialog: boolean,
    dialogName: string
  ) {
    const rowData = info.selectedObject;
    if (!mapping || !rowData || !rowData?.has(mapping['entityID'])) {
      return;
    }
    const cardID = rowData.get(mapping['entityID']) as string;
    if (!cardID || Guid.isEmpty(cardID) || !Guid.isValid(cardID)) {
      return;
    }
    let cardName = '';
    if (mapping['entityName']) {
      cardName = rowData.get(mapping['entityName']) as string;
    }
    let caption = cardName;
    if (!caption) {
      caption = cardID;
    }
    const viewControlInfo = new CardViewControlInfo();
    viewControlInfo.id = cardID;
    viewControlInfo.displayText = caption;
    viewControlInfo.controlName = mapping['controlName'];
    viewControlInfo.viewAlias = mapping['viewName'];
    viewControlInfo.colPrefix = mapping['colPrefix'];
    const requestInfo = {};
    viewControlInfo.setInfo(requestInfo);

    const uiContextInstance = UIContext.create(info.context);
    try {
      if (displayInDialog) {
        await showLoadingOverlay(async () => {
          await AdvancedCardDialogManager.instance.openCard({
            cardId: cardID,
            displayValue: caption,
            info: requestInfo,
            context: info.context,
            dialogOptions: {
              dialogName
            }
          });
        });
      } else {
        await showLoadingOverlay(async () => {
          await openCard({
            cardId: cardID,
            displayValue: caption,
            info: requestInfo,
            context: info.context
          });
        });
      }
    } finally {
      uiContextInstance.dispose();
    }
  }

  private static tryGetCardReferenceMapping(
    viewModel: ViewControlViewModel,
    prefixReference: string
  ) {
    const viewMetadata = viewModel.viewMetadata;
    if (!viewMetadata) {
      return null;
    }
    if (prefixReference) {
      const reference = viewMetadata.references.get(prefixReference);
      if (!reference) {
        return null;
      }
      const mapping = OpenCardInViewUIExtension.tryGetCardIDNameMapping(
        viewMetadata,
        reference!.colPrefix,
        reference!.displayValueColumn
      );
      return OpenCardInViewUIExtension.tryAttachControlInfo(mapping, viewModel);
    }

    for (const reference of viewMetadata.references.values()) {
      if (!reference.isCard || !reference.openOnDoubleClick) {
        continue;
      }
      const mapping = OpenCardInViewUIExtension.tryGetCardIDNameMapping(
        viewMetadata,
        reference.colPrefix,
        reference.displayValueColumn
      );
      if (mapping) {
        return OpenCardInViewUIExtension.tryAttachControlInfo(mapping, viewModel);
      }
    }
    return null;
  }

  private static tryGetCardIDNameMapping(
    viewMetadata: ViewMetadataSealed | null,
    prefixReference: string | null,
    displayValueColumn: string | null = null
  ) {
    if (!viewMetadata || !prefixReference) {
      return null;
    }
    // попытка найти колонку с идентификатором
    const column =
      viewMetadata.columns.get(prefixReference + 'ID') ??
      viewMetadata.columns.get(prefixReference + 'RowID');

    if (!column) {
      return null;
    }

    const mapping = { entityID: column.alias, colPrefix: prefixReference };
    if (!displayValueColumn) {
      for (const [key, value] of viewMetadata.columns) {
        if (!key.startsWith(prefixReference)) {
          continue;
        }
        if (value == column) {
          continue;
        }
        displayValueColumn = key;
        break;
      }
    }
    mapping['entityName'] = displayValueColumn;
    return mapping;
  }

  private static tryAttachControlInfo(mapping, viewModel: ViewControlViewModel) {
    if (mapping) {
      mapping['controlName'] = viewModel.cardTypeControl.name;
      mapping['viewName'] = viewModel.viewMetadata?.alias;
    }
    return mapping;
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/OpenCardInViewUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Cards;
using Tessa.Localization;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Controls;
using Tessa.Views.Metadata;

namespace Tessa.Extensions.Default.Client.UI
{
    /// <summary>
    /// Реализация расширения типа карточки для добавления возможности
    /// открывать карточки из представления.
    /// </summary>
    public sealed class OpenCardInViewUIExtension : CardUIExtension
    {
        #region Fields

        private readonly ICardMetadata cardMetadata;
        private readonly IUIHost uiHost;
        private readonly IAdvancedCardDialogManager cardDialogManager;

        #endregion

        #region Constructors

        public OpenCardInViewUIExtension(
            ICardMetadata cardMetadata,
            IUIHost uiHost,
            IAdvancedCardDialogManager cardDialogManager)
        {
            this.cardMetadata = cardMetadata ?? throw new ArgumentNullException(nameof(cardMetadata));
            this.uiHost = uiHost ?? throw new ArgumentNullException(nameof(uiHost));
            this.cardDialogManager = cardDialogManager ?? throw new ArgumentNullException(nameof(cardDialogManager));
        }

        #endregion

        #region Type extension methods

        /// <summary>
        /// Помощник применения функционала данного расширения.
        /// </summary>
        /// <param name="context">Контекст расширения типа.</param>
        /// <returns>Ассинхронная задача.</returns>
        private async Task ExecuteInitializedAsync(ITypeExtensionContext context)
        {
            var extensionContext = (ICardUIExtensionContext) context.ExternalContext;
            if (context.CardTask is null)
            {
                // действия для карточки
                await this.AttachDoubleClickHandlerAsync(
                    extensionContext.UIContext,
                    extensionContext.Model,
                    context.Settings,
                    cancellationToken: context.CancellationToken);
            }
            else
            {
                // действия для задания
                await extensionContext.Model.ModifyTasksAsync(async (task, model) =>
                {
                    if (task.TaskModel.CardTask == context.CardTask)
                    {
                        await task.ModifyWorkspaceAsync(async (t, subscribeToTaskModel) =>
                        {
                            await this.AttachDoubleClickHandlerAsync(
                                extensionContext.UIContext,
                                task.TaskModel,
                                context.Settings);
                        });
                    }
                });
            }
        }

        #endregion

        #region Private methods

        /// <summary>
        /// Помощник подключения функционала открытия карточки по двойному клику.
        /// </summary>
        /// <param name="uiContext">Контекст операции с пользовательским интерфейсом.</param>
        /// <param name="cardModel">Модель карточки.</param>
        /// <param name="settings">Настройки расширения.</param>
        /// <param name="cancellationToken">Объект для отмены асинхронной операции.</param>
        /// <returns>Асинхронная задача с результатом успешности присоединения обработчика.</returns>
        private async Task<bool> AttachDoubleClickHandlerAsync(
            IUIContext uiContext,
            ICardModel cardModel,
            ISerializableObject settings,
            CancellationToken cancellationToken = default)
        {
            // получаем имя представления
            var viewControlName = settings.TryGet<string>(DefaultCardTypeExtensionSettings.ViewControlAlias);

            // пытаемся получить контрол по имени
            var viewModel = cardModel.Controls.TryGet<CardViewControlViewModel>(viewControlName);
            if (viewModel is null)
            {
                return false;
            }
            
            // определяем колонку с идентификатором карточки
            var prefixReference = settings.TryGet<string>(DefaultCardTypeExtensionSettings.ViewReferencePrefix)?.Trim();
            var mapping = TryGetCardReferenceMapping(viewModel, prefixReference);
            var displayInDialog = settings.TryGet<bool>(DefaultCardTypeExtensionSettings.IsOpenCardInDialog);
            var dialogName = settings.TryGet<string>(DefaultCardTypeExtensionSettings.CardDialogName);

            // присоединяем наш обработчик
            viewModel.DoubleClickCommand = new DelegateCommand(async p =>
            {
                if (!string.IsNullOrEmpty(prefixReference) && mapping is null)
                {
                    string message = LocalizationManager.Format(
                        "$UI_Cards_TypesEditor_Exception_RefSectionInView",
                        prefixReference,
                        viewModel.CardTypeControl.ToString(),
                        viewModel.Block.CardTypeBlock.ToString(),
                        viewModel.Block.Form.CardTypeForm.ToString(),
                        viewModel.CardModel.CardType.ToString());
                    throw new InvalidOperationException(message);
                }
                await this.DoubleClickHandlerAsync(
                    uiContext,
                    (IViewClickInfo) p,
                    mapping,
                    displayInDialog,
                    dialogName);
            });

            return true;
        }

        /// <summary>
        /// Обработчик двойного клика мыши.
        /// </summary>
        /// <param name="uiContext">Контекст операции с пользовательским интерфейсом.</param>
        /// <param name="clickInfo">Информация о текущей выделенной в представлении строке.</param>
        /// <param name="mapping">Названия колонок с идентификатором и именем.</param>
        /// <param name="displayInDialog">Флаг необходимости открывать карточку в диалоговом режиме.</param>
        /// <param name="dialogName">Служебное имя диалога.</param>
        /// <returns>Задача.</returns>
        private async Task DoubleClickHandlerAsync(
            IUIContext uiContext,
            IViewClickInfo clickInfo,
            ViewMapping mapping,
            bool displayInDialog,
            string dialogName)
        {
            var row = clickInfo.Row;
            if (mapping is null ||
                row.Data[mapping.EntityID] is not Guid cardID ||
                cardID == Guid.Empty)
            {
                return;
            }
            
            string cardName = null;
            if (!string.IsNullOrWhiteSpace(mapping.EntityName))
            {
                cardName = row.Data[mapping.EntityName] as string;
            }
            var caption = cardName;
            if (string.IsNullOrWhiteSpace(caption))
            {
                caption = cardID.ToString();
            }

            var viewControlInfo = new CardViewControlInfo
            {
                ID = cardID,
                DisplayText = caption,
                ControlName = mapping.ControlName,
                ViewAlias = mapping.ViewName,
                ColPrefix = mapping.ColPrefix
            };
            var info = new Dictionary<string, object>(StringComparer.Ordinal);
            viewControlInfo.Set(info);

            using ISplash splash = TessaSplash.Create(TessaSplashMessage.OpeningCard);
            using var _ = UIContext.Create(uiContext);
            try
            {
                // теперь проверяем режим открытия
                if (displayInDialog)
                {
                    // открывать в диалоге
                    await this.cardDialogManager.OpenCardAsync(
                        cardID,
                        dialogName: dialogName,
                        options: new OpenCardOptions
                        {
                            DisplayValue = caption,
                            Info = info,
                            UIContext = uiContext,
                            Splash = splash,
                        });
                }
                else
                {
                    // открывать во вкладке или диалоге, если карточка уже открыта в диалоге
                    await this.uiHost.OpenCardAsync(
                        cardID,
                        options: new OpenCardOptions
                        {
                            DisplayValue = caption,
                            Info = info,
                            UIContext = uiContext,
                            Splash = splash,
                        });
                }
            }
            catch (NotSupportedException)
            {
                // используется FakeUIHost, например, мы открыты в TessaAdmin, игнорируем ошибку
            }
        }

        /// <summary>
        /// Помощник получения маппинга колонок представления.
        /// </summary>
        /// <param name="viewModel">Модель контрола представления.</param>
        /// <param name="prefixReference">Префикс целевых колонок представления.</param>
        /// <returns>Возвращает маппинг или <c>null</c> - если маппинг не найден.</returns>
        private static ViewMapping TryGetCardReferenceMapping(
            CardViewControlViewModel viewModel,
            string prefixReference)
        {
            var viewMetadata = viewModel.ViewMetadata;
            if (viewMetadata is null)
            {
                return null;
            }

            if (!string.IsNullOrEmpty(prefixReference))
            {
                // пытаемся найти нужную ссылку
                var reference = viewMetadata.References.FindByName(prefixReference);
                if (reference is null)
                {
                    return null;
                }
                var mapping = TryGetCardIDNameMapping(viewMetadata, reference.ColPrefix, reference.DisplayValueColumn);
                return TryAttachControlInfo(mapping, viewModel);
            }
            
            foreach (var reference in viewMetadata.References)
            {
                if (!reference.IsCard || !reference.OpenOnDoubleClick)
                {
                    continue;
                }
                var mapping = TryGetCardIDNameMapping(viewMetadata, reference.ColPrefix, reference.DisplayValueColumn);
                if (mapping is not null)
                {
                    return TryAttachControlInfo(mapping, viewModel);
                }
            }

            return null;
        }

        /// <summary>
        /// Помощник получения маппинга колонок представления.
        /// </summary>
        /// <param name="viewMetadata">Метаданные представления.</param>
        /// <param name="prefixReference">Префикс целевых колонок представления.</param>
        /// <param name="displayValueColumn">Имя колонки с отображаемым знанчением. Может быть <c>null</c>.</param>
        /// <returns>Возвращает маппинг или <c>null</c> - если маппинг не найден.</returns>
        private static ViewMapping TryGetCardIDNameMapping(
            IViewMetadata viewMetadata,
            string prefixReference,
            string displayValueColumn = null)
        {
            if (viewMetadata is null ||
                string.IsNullOrEmpty(prefixReference))
            {
                return null;
            }
                      
            var column = viewMetadata.Columns.FindByName(prefixReference + "ID") ??
                viewMetadata.Columns.FindByName(prefixReference + "RowID");
            if (column is null)
            {
                return null;
            }

            var mapping = new ViewMapping()
            {
                ColPrefix = prefixReference,
                EntityID = column.Alias
            };

            if (string.IsNullOrEmpty(displayValueColumn))
            {
                displayValueColumn = viewMetadata.Columns
                    .Where(c => c.Alias.StartsWith(prefixReference, StringComparison.OrdinalIgnoreCase) &&
                                c != column)
                    .Select(c => c.Alias)
                    .FirstOrDefault();
            }
            mapping.EntityName = displayValueColumn;

            return mapping;
        }

        /// <summary>
        /// Помощник присоединения данных о контроле к маппингу колонок представления.
        /// </summary>
        /// <param name="mapping">Маппинг колонок представления.</param>
        /// <param name="viewModel">Модель представления.</param>
        /// <returns>Маппинг колонок представления с присоединённой информацией.</returns>
        private static ViewMapping TryAttachControlInfo(
            ViewMapping mapping,
            CardViewControlViewModel viewModel)
        {
            if (mapping is not null)
            {
                mapping.ControlName = viewModel.CardTypeControl.Name;
                mapping.ViewName = viewModel.Alias;
            }
            return mapping;
        }

        #endregion

        #region Base overrides

        /// <inheritdoc/>
        public override async Task Initialized(ICardUIExtensionContext context)
        {
            ValidationResult result = await CardHelper
                .ExecuteTypeExtensionsAsync(
                    DefaultCardTypeExtensionTypes.OpenCardInView,
                    context.Card,
                    this.cardMetadata,
                    ExecuteInitializedAsync,
                    context,
                    cancellationToken: context.CancellationToken);

            context.ValidationResult.Add(result);
        }

        #endregion

        #region Internal Classes

        /// <summary>
        /// Класс для маппинга колонок в представлении.
        /// </summary>
        class ViewMapping
        {
            /// <summary>
            /// Имя элемента управления.
            /// </summary>
            internal string ControlName;
            /// <summary>
            /// Имя представления.
            /// </summary>
            internal string ViewName;
            /// <summary>
            /// Префикс референса.
            /// </summary>
            internal string ColPrefix;
            /// <summary>
            /// Название колонки с идентификатором.
            /// </summary>
            internal string EntityID;
            /// <summary>
            /// Название колонки с именем.
            /// </summary>
            internal string EntityName;
        }

        #endregion
    }
}
